Userland/Demos/CatDog/CatDog.h

```diff
@@ -25,112 +25,39 @@ class CatDog final : public GUI::Widget
-    {
-        return state == MainState::Idle || state == MainState::Alerted || state == MainState::Sleeping;
-    }
-
-    virtual void timer_event(Core::TimerEvent&) override;
-    virtual void paint_event(GUI::PaintEvent& event) override;
-    virtual void track_mouse_move(Gfx::IntPoint point) override;
-    virtual void mousedown_event(GUI::MouseEvent& event) override;
-    virtual void context_menu_event(GUI::ContextMenuEvent& event) override;
-
-    void start_the_timer() { m_timer.start(); }
+    void timer_event(Core::TimerEvent&) final;
+    void paint_event(GUI::PaintEvent& event) final;
+    void track_mouse_move(Gfx::IntPoint point) final;
+    void mousedown_event(GUI::MouseEvent& event) final;
+    void context_menu_event(GUI::ContextMenuEvent& event) final;
 
     Function<void()> on_click;
     Function<void(GUI::ContextMenuEvent&)> on_context_menu_request;
 
-    bool roaming() const { return m_roaming; }
-    void set_roaming(bool roaming)
-    {
-        m_roaming = roaming;
-        if (!roaming) {
-            // If we stop CatDog while it's in a program-specific state, we don't want it to be alerted.
-            if (m_main_state == MainState::Idle || m_main_state == MainState::Sleeping)
-                m_main_state = MainState::Alerted;
-            m_curr_frame = 0;
-            set_image_by_main_state();
-            update();
-        }
-    }
+    void set_roaming(bool roaming);
 
-    MainState main_state() const { return m_main_state; }
+    [[nodiscard]] bool is_artist() const;
+    [[nodiscard]] bool is_inspector() const;
 
 private:
-    Gfx::IntPoint m_temp_pos;
-    Core::ElapsedTimer m_timer;
+    struct ImageForState {
+        uint16_t state;
+        RefPtr<Gfx::Bitmap> bitmap;
+    };
 
-    int m_curr_frame = 1;
-    int m_moveX, m_moveY = 0;
-    MainState m_main_state { MainState::Alerted };
-    bool m_up, m_down, m_left, m_right;
-    bool m_roaming { true };
+    Vector<ImageForState> m_images;
 
-    NonnullOwnPtr<Core::Stream::File> m_proc_all;
+    Gfx::IntPoint m_mouse_offset {};
+    Core::ElapsedTimer m_idle_sleep_timer;
 
-    RefPtr<Gfx::Bitmap> m_alert;
-    RefPtr<Gfx::Bitmap> m_artist;
-    RefPtr<Gfx::Bitmap> m_erun1;
-    RefPtr<Gfx::Bitmap> m_erun2;
-    RefPtr<Gfx::Bitmap> m_inspector;
-    RefPtr<Gfx::Bitmap> m_nerun1;
-    RefPtr<Gfx::Bitmap> m_nerun2;
-    RefPtr<Gfx::Bitmap> m_nrun1;
-    RefPtr<Gfx::Bitmap> m_nrun2;
-    RefPtr<Gfx::Bitmap> m_nwrun1;
-    RefPtr<Gfx::Bitmap> m_nwrun2;
-    RefPtr<Gfx::Bitmap> m_serun1;
-    RefPtr<Gfx::Bitmap> m_serun2;
-    RefPtr<Gfx::Bitmap> m_sleep1;
-    RefPtr<Gfx::Bitmap> m_sleep2;
-    RefPtr<Gfx::Bitmap> m_srun1;
-    RefPtr<Gfx::Bitmap> m_srun2;
-    RefPtr<Gfx::Bitmap> m_still;
-    RefPtr<Gfx::Bitmap> m_swrun1;
-    RefPtr<Gfx::Bitmap> m_swrun2;
-    RefPtr<Gfx::Bitmap> m_wrun1;
-    RefPtr<Gfx::Bitmap> m_wrun2;
+    NonnullOwnPtr<Core::Stream::File> m_proc_all;
 
-    RefPtr<Gfx::Bitmap> m_curr_bmp;
+    uint16_t m_state {};
```

Comment:
Could we make at least the initialization of `m_state` a bit more explicit by stating this starts as `Alert`?

Reply:
Can't do it there without unnecessarily exposing the State enum in the header file, but I did it in the constructor. Done.